### DISCUSSION TITLE
Advanced graphiql

### DISCUSSION THREAD

**Discussion by manuelwehrli** [directus/directus#4809] [Feature Requests]
Instead of the standard graphiql on the graphql endpoint, this advanced graphiql would offer better DX: https://github.com/hasura/graphiql-online It has the a nice header editor and a practical api explorer where you can build queries quickly just by clicking. Can we have this integrated into directus?

**Comment by rijkvanzanten** [comment]
I'm not sure if shipping a whole copy of GraphiQL in production Directus makes sense to begin with 🤔  In any case, you should be able to point https://graphiql-online.com to your Directus instance 👍🏻

**Reply by manuelwehrli** [reply]
Of course, we can just use it as a separate tool, the question is to replace the currently integrated graphiql, for development mode at least. As for now, directus comes with one graphiql anyway, so why not use the most powerful one?

**Reply by rijkvanzanten** [reply]
The one included was the one included from `express-graphql`, it wasn't a custom included GraphiQL instance that we can just swap out. https://github.com/directus/directus/pull/4625 removes the reliance on graphql, which opens up the ability to implement this idea in the future

**Reply by beebase** [reply]
Is there an example how to do that?

> I'm not sure if shipping a whole copy of GraphiQL in production Directus makes sense to begin with 🤔 In any case, you should be able to point https://graphiql-online.com/ to your Directus instance

**Comment by ravi-2912** [comment]
When I wanted GraphiQL with Directus, I too saw this discussion. However, I found an alternative - I created a custom endpoint within Directus and added GraphQL Playground (which is now Apollo's official) to it. Here's my code below (don't forget to `npm install graphql-playground-middleware-express` to Directus installation or container).

```javascript
const helmet = require('helmet');
const expressPlayground = require('graphql-playground-middleware-express').default;

export default (router) => {
    router.use(
        '/',
        helmet({
            contentSecurityPolicy: {
                directives: {
                    'script-src': ["'self'", "'unsafe-inline'", 'cdn.jsdelivr.net'],
                    'img-src': ["'self'", 'data:', 'cdn.jsdelivr.net'],
                    'style-src': ["'self'", 'https:', "'unsafe-inline'", 'cdn.jsdelivr.net'],
                    'font-src': ["'self'", 'https:', 'data:', 'cdn.jsdelivr.net'],
                    'default-src': ["'self'", "'unsafe-inline'", 'cdn.jsdelivr.net'],
                },
            },
        }),
        expressPlayground({
            endpoint: '/graphql',
        })
    );
};
```

All I now need to do is add authentication to this endpoint, which I am stil exploring.

**Reply by manuelwehrli** [reply]
Though this is a nice idea, I think it makes only sense to host a visual graphql tool for a public api used by many different actors so they can quickly test and fiddle around with it. For a more private api, I recommend that every developer uses his own local tools which most of the time are more powerful. This discussion originally came up when directus had graphiql built-in which was removed afterwards.

**Reply by ravi-2912** [reply]
True, for public APIs I would do that. I had a different use case. Thanks for your inputs. I only posted this solution here not for GraphiQL but actually for how to modify the Content Security Policies (CSP) in `helmet`. I would appreciate any recommendations on wether this way to modify CSP is good or not?

**Reply by rijkvanzanten** [reply]
> I would appreciate any recommendations on wether this way to modify CSP is good or not?

I'd recommend using the environment variables for `CONTENT_SECURITY_POLICY_*` instead, like `CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC="'self','unsafe-inline',cdn.jsdelivr.net"`

**Reply by ravi-2912** [reply]
> 

But will this not add the CSPs to all custom endpoints?

**Reply by rijkvanzanten** [reply]
It will add the CSP to all endpoints, yes

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
